<launch>

  <arg name="camera_name" default=""/>
  <arg name="camera_id" default="0"/>
  <arg name="stream_role" default="still"/>
  <arg name="pixel_format" default="RGB888"/>
  <arg name="frame_id" default="libcamera_frame"/>
  <arg name="calib_url" default="file://$(find libcamera_ros)/config/calib/libcamera.yaml"/>
  <arg name="resolution_width" default="1456"/>
  <arg name="resolution_height" default="1088"/>

  <!-- will it run using GNU debugger? -->
  <arg name="DEBUG" default="false" />
  <arg unless="$(arg DEBUG)" name="launch_prefix_debug" value=""/>
  <arg     if="$(arg DEBUG)" name="launch_prefix_debug" value="debug_roslaunch"/>

  <!-- will it run as standalone nodelet or using a nodelet manager? -->
  <arg name="standalone" default="true" />
  <arg name="manager" default="libcamera_ros_manager" />

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg manager)"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <env name="LIBPISP_BE_CONFIG_FILE" value="$(optenv LIBPISP_BE_CONFIG_FILE /opt/ros/noetic/share/libpisp/backend_default_config.json)" />
  <env name="LIBCAMERA_IPA_MODULE_PATH" value="$(optenv LIBCAMERA_IPA_MODULE_PATH /opt/ros/noetic/lib/libcamera/)" />
  <env name="LIBCAMERA_IPA_CONFIG_PATH" value="$(optenv LIBCAMERA_IPA_CONFIG_PATH /opt/ros/noetic/share/libcamera/ipa)" />
  <env if="$(arg DEBUG)" name="LIBCAMERA_LOG_LEVELS" value="$(optenv LIBCAMERA_LOG_LEVELS *:DEBUG)" />

  <node pkg="nodelet" type="nodelet" name="libcamera_ros" args="$(arg nodelet) libcamera_ros/LibcameraRos $(arg nodelet_manager)" launch-prefix="$(arg launch_prefix_debug)" output="screen">

    <param name="camera_name" type="string" value="$(arg camera_name)"/>
    <param name="camera_id" type="int" value="$(arg camera_id)"/>
    <param name="stream_role" type="string" value="$(arg stream_role)"/>
    <param name="pixel_format" type="string" value="$(arg pixel_format)"/>
    <param name="frame_id" type="string" value="$(arg frame_id)"/>
    <param name="calib_url" type="string" value="$(arg calib_url)"/>
    <param name="resolution/width" type="int" value="$(arg resolution_width)"/>
    <param name="resolution/height" type="int" value="$(arg resolution_height)"/>

    <!-- Disable publishing of specific image transport topics -->
    <rosparam param="image_raw/disable_pub_plugins">
    - 'image_transport/compressedDepth'
    - 'image_transport/theora'
    - 'image_transport/h264'
    </rosparam>

    <!-- Publishers -->
    <remap from="~image_raw" to="~image_raw" />
    <remap from="~camera_info" to="~camera_info" />

  </node>

</launch>
